### ISSUE TITLE
Bucket file layout

### ISSUE THREAD

**Issue by jlebon** [coreos/fedora-coreos-tracker#189]
How should our various files (metadata & build artifacts) be laid out in the bucket? Note this isn't something that we expect users to care about. Some pieces of published metadata will hold links to files into the bucket, but we should be able to change the bucket layout itself without impacting users. As such, we should also strongly discourage perusing the bucket (which without indexing it would be cumbersome anyway) or predicting locations for new builds.

Here's a strawman based on some initial thoughts in https://github.com/coreos/coreos-assembler/issues/159:

```
/
    prod/
        streams/
            stable/
                stream.json                   # holds stream metadata
                builds/
                    builds.json               # raw cosa builds listing metadata
                    30.1234-5/
                        release.json          # holds release metadata
                        x86_64/
                            meta.json         # raw cosa build metadata
                            commitmeta.json
                            fedora-coreos-30.8-qemu.qcow2.gz
                            ostree-commit-object
                            ostree-commit.tar
                            ...
                        ppc64le/
                            ...
                        ...
            testing/
            next/
            ...
```

**Comment by dustymabe** [comment]
looks mostly good a few comments:

- i'd prefer to have filename uniqueness so maybe let's not have two `meta.json` filenames
- could we add architecture awareness to the example in the description?

**Comment by jlebon** [comment]
> i'd prefer to have filename uniqueness so maybe let's not have two `meta.json` filenames

Yeah fair. Maybe just `stream.json` in that case?

> could we add architecture awareness to the example in the description?

This is mostly dependent on https://github.com/coreos/coreos-assembler/issues/463; though I guess how the various archful builds are laid out are related but somewhat independent of how they get there, so we could strawman that part now too.

**Comment by dustymabe** [comment]
> so we could strawman that part now too.

Yeah - a concrete example (like the one in your description) is the best way I can think to actually hash it out. My proposal is we simply add arch directories under the buildid:

```
        builds/
          30.1234-5/      # holds COSA build artifacts
                x86_64/
                aarch64/
                ppc64le/ 

```

**Comment by dustymabe** [comment]
> Yeah fair. Maybe just `stream.json` in that case?

forgot to respond to this: 👍

**Comment by jlebon** [comment]
OK, updated initial comment with those modifications!

**Comment by arithx** [comment]
Couple of questions I need answered for the `plume` work:

1) Is it intended that there is a different bucket location that these artifacts are initially written to and then copied to this structure via `plume` in the release process?
2) Where are we storing the release metadata index in this structure (right under `prod/`)?
3) Are all files/artifacts inside of the individual build level buckets things that are meant to be publicly readable (e.x.: should `meta.json` be fetchable by end users)?

**Comment by bgilbert** [comment]
3. Yes.

Note that if the `release.json` is staged somewhere by COSA and then copied over by plume, incremental updates to it (e.g. to backfill an AMI ID) will need to be copied over again.

**Comment by sinnykumari** [comment]
Few things need to understand to work on [Stream Metadata Generator](https://github.com/coreos/fedora-coreos-tracker/issues/193):
1. Is stream.json mentioned in design is where Stream Metadata Generator output will be uploaded and made available? 
2. release metadata (release.json ) is available for each builds in a stream. Which one of them stream metadata generator will be using as input for building stream metadata? Is there something like latest/ under builds/ directory pointing to the latest one to be utilized?

**Comment by jlebon** [comment]
> Is stream.json mentioned in design is where Stream Metadata Generator output will be uploaded and made available?

Yes, I think that's correct.

> Is there something like latest/ under builds/ directory pointing to the latest one to be utilized?

Whoops, I had missed the `builds.json` here from COSA. I updated the first comment with that. So e.g. you can determine what the latest build is by looking at `builds.json`.

**Comment by sinnykumari** [comment]
Makes sense, thanks @jlebon

**Comment by bgilbert** [comment]
The current plan is not to use `builds.json` for that, but instead to use a separate JSON document (https://github.com/coreos/fedora-coreos-tracker/issues/98#issuecomment-497521551) pointing to the release metadata of released builds.

**Comment by jlebon** [comment]
Ahh yeah, right. The `builds.json` is essentially a raw listing of all the cosa builds stored in that directory. Whereas the "release index" (do we another name for this yet?) actually represents releases we want users to consume.

> 2. Where are we storing the release metadata index in this structure (right under `prod/`)?

Hmm, seems like this should be a per-stream thing, right? So maybe at `/prod/streams/$stream/releases.json` ? (Essentially sitting side by side with `stream.json`).

**Comment by bgilbert** [comment]
We don't have another name.  "Release index" works for me.  And yes, per-stream.

**Comment by arithx** [comment]
@jlebon For the individual build release metadata right now the diagram seems to indicate one per architecture, is this correct? Or should there be one per build version (e.x.: `prod/<channel>/builds/<version>/release.json`)?

**Comment by sinnykumari** [comment]
> The current plan is not to use `builds.json` for that, but instead to use a separate JSON document ([#98 (comment)](https://github.com/coreos/fedora-coreos-tracker/issues/98#issuecomment-497521551)) pointing to the release metadata of released builds.

By looking at json layout for release index at https://github.com/coreos/fedora-coreos-tracker/issues/98#issuecomment-497521551 ,  how does one know rwhich release metadata: <endpount> is the latest one? Not sure if having OSTree version is enough to find that information.

**Comment by jlebon** [comment]
> For the individual build release metadata right now the diagram seems to indicate one per architecture, is this correct? Or should there be one per build version (e.x.: `prod/<channel>/builds/<version>/release.json`)?

Yeah good point, I think that's indeed what we want so that it's a single file representing the whole build. Amended first comment!

So thinking more on this, in https://github.com/coreos/fedora-coreos-tracker/issues/98#issuecomment-496774211 regarding `release.json`:

> This seems largely redundant with `meta.json`, so in principle we might skip it. But it could also be a good opportunity to clean up that format a bit.

I think I would agree with this. Let `meta.json` be the raw cosa output, and `release.json` only have a filtered (and cross-arch collected) version of it so we don't end up exposing more information than we need to (e.g. offhand, the `pkgdiff` is not really useful; diffs should be done at a higher level where we actually know what we're releasing).

**Comment by arithx** [comment]
> Is it intended that there is a different bucket location that these artifacts are initially written to and then copied to this structure via plume in the release process?

From discussions w/ @jlebon & @bgilbert  we're going to move forward with a single bucket and just modify the object permissions to be publicly readable as part of the release process.

**Comment by bgilbert** [comment]
> By looking at json layout for release index at https://github.com/coreos/fedora-coreos-tracker/issues/98#issuecomment-497521551, how does one know rwhich release metadata: is the latest one?

Everything except the stream metadata builder should be getting that info from the stream metadata.  The metadata builder could look at:

1. A timestamp for each record, if we added one;
2. The highest version number; or
3. The last release in the list, if we always appended releases to the end.

3 seems easiest, and I don't think it has substantial downsides?

**Comment by sinnykumari** [comment]
I like option 3 as well, don't see any downside if we keep convention that latest release metadata info gets always appended in the index file of release metadata.

**Comment by jlebon** [comment]
Officially proposed in https://github.com/coreos/fedora-coreos-tracker/pull/208. (Not that we can't make tweaks later on of course).

**Comment by bgilbert** [comment]
Discussed OOB: we should move the stream metadata to a separate directory (e.g. `/streams/stable.json`) so we can document its URL as the canonical location for stream metadata.

**Comment by jlebon** [comment]
Resolved in #208.
